### ISSUE TITLE
Question about DeepCpG

### ISSUE THREAD

**Issue by daquang** [cangermueller/deepcpg#3]
Hi, Daniel, author of DanQ here. Thanks for citing my paper in your work. I have a quick question about DeepCpG's inputs. Are you using CpG state labels from neighboring sites as inputs to your recurrent neural network? If so, your method sort of blurs the line between supervised and unsupervised learning in the sense that you are using supervised learning methods to train a model that uses CpG states as both labels and inputs. 

Also, in your paper, you said in the CpG module each xt is a vector of size 100 containing the "methylation state and distance of K = 25 CpG sites to the left and to the right of a target CpG". From what I understand, xt should contain information on 50 total CpG sites, and is assigning 2 binary features for each of the 50 sites, bringing the total to 100. If so, is there a reason why you could not have brought this down to just 50 features and assigned a single value to each of the 50 sites? I always assumed that methylation was completely binary (either yes or no), but I might be confused here.

Thanks,
Daniel

**Comment by daquang** [comment]
Sorry, I think I just understoof what you meant by the 100-size vectors. You have 50 CpG sites as inputs, and each site has a methylation value and a distance value. Am I correct in assuming these methylation values are also labels for other training samples?

**Comment by cangermueller** [comment]
Hi Daniel,

yes, this is correct: 100 features since every CpG site has a state and a distance. DeepCpG is not unsupervised since it does not use predicted CpG neighbors as input. But you are right, one might also train a generative recurrent network for modeling the probability P(x1, ..., xN) of a sequence of methylation states x1, ..., xN, akin to [generating handwriting](https://arxiv.org/abs/1308.0850).

**Comment by daquang** [comment]
Thanks for clarifying. 

What I meant to ask was, do the same numerical values that appear as CpG state inputs also appear as CpG state outputs? For example, let's say you have two adjacent CpG sites: n and n-1. Would the CpG state at position n-1 appear as an input for CpG site n, where the output would be the CpG state at site n? And vice versa: Would the CpG state at position n appear as an input for CpG site n-1, where the output would be the CpG state at site n-1?

**Comment by cangermueller** [comment]
No, DeepCpG uses as input only **observed** neighboring CpG sites, not model outputs.

**Comment by daquang** [comment]
Ah, I see. So, you have a set of CpG sites that are exclusively used as target labels, making sure their CpG states never appear as inputs to the model for other training samples. It wasn't quite clear in the paper.

**Comment by dnaase** [comment]
Sorry, I am not quite sure if I understand the model correctly. If the input is represented as the following graph: (M is missing cpg, 
 X11, X12, X13 ... means the observed cpg, which have reads covered there;  T1,T2 means the cpg you would like to predict by the model)

cell1:    M-M-M-X11-M-M-T1-M-X12-X13-M-M
cell2:   M-X21-M-M-M-M-T2-X22-M-M-M-X23
...

If your model want neighbor 1 CpG, do you mean the input is like this way?

for cell 1:  X11,X12
for cell 2: X21, X22

Or:
for cell 1:  M,M
for cell 2: M, X22
(and M is substituted to some value..)

thanks!

**Comment by cangermueller** [comment]
The input will be
for cell 1: X11,X12
for cell 2: X21, X22
